lib/core/cmap/message_stream.js

```diff
@@ -0,0 +1,177 @@
+          new MongoParseError(
+            `Invalid message size: ${sizeOfMessage}, max allowed: ${this.maxBsonMessageSize}`
+          )
+        );
+        return;
+      }
+
+      if (sizeOfMessage > this[kBuffer].length) {
+        callback();
+        return;
+      }
+
+      const messageBuffer = this[kBuffer].slice(0, sizeOfMessage);
+      processMessage(this, messageBuffer, callback);
+      this[kBuffer].consume(sizeOfMessage);
+    }
+  }
+
+  _read(/* size */) {
+    // NOTE: This implementation is empty because we explicitly push data to be read
+    //       when `writeMessage` is called.
+    return;
+  }
+
+  writeCommand(command, operationDescription) {
+    // TODO: agreed compressor should live in `StreamDescription`
+    const shouldCompress = operationDescription && !!operationDescription.agreedCompressor;
+    if (!shouldCompress || !canCompress(command)) {
+      this.push(Buffer.concat(command.toBin()));
+      return;
+    }
+
+    // otherwise, compress the message
+    const concatenatedOriginalCommandBuffer = Buffer.concat(command.toBind());
+    const messageToBeCompressed = concatenatedOriginalCommandBuffer.slice(MESSAGE_HEADER_SIZE);
+
+    // Extract information needed for OP_COMPRESSED from the uncompressed message
+    const originalCommandOpCode = concatenatedOriginalCommandBuffer.readInt32LE(12);
+
+    // Compress the message body
+    compress({ options: operationDescription }, messageToBeCompressed, (err, compressedMessage) => {
+      if (err) {
+        operationDescription.cb(err, null);
+        return;
+      }
+
+      // Create the msgHeader of OP_COMPRESSED
+      const msgHeader = Buffer.alloc(MESSAGE_HEADER_SIZE);
+      msgHeader.writeInt32LE(
+        MESSAGE_HEADER_SIZE + COMPRESSION_DETAILS_SIZE + compressedMessage.length,
+        0
+      ); // messageLength
+      msgHeader.writeInt32LE(command.requestId, 4); // requestID
+      msgHeader.writeInt32LE(0, 8); // responseTo (zero)
+      msgHeader.writeInt32LE(opcodes.OP_COMPRESSED, 12); // opCode
+
+      // Create the compression details of OP_COMPRESSED
+      const compressionDetails = Buffer.alloc(COMPRESSION_DETAILS_SIZE);
+      compressionDetails.writeInt32LE(originalCommandOpCode, 0); // originalOpcode
+      compressionDetails.writeInt32LE(messageToBeCompressed.length, 4); // Size of the uncompressed compressedMessage, excluding the MsgHeader
+      compressionDetails.writeUInt8(compressorIDs[operationDescription.agreedCompressor], 8); // compressorID
+
+      this.push(Buffer.concat([msgHeader, compressionDetails, compressedMessage]));
+    });
+  }
+}
+
+// Return whether a command contains an uncompressible command term
+// Will return true if command contains no uncompressible command terms
+function canCompress(command) {
+  const commandDoc = command instanceof Msg ? command.command : command.query;
+  const commandName = Object.keys(commandDoc)[0];
+  return uncompressibleCommands.indexOf(commandName) === -1;
```

Comment:
`Set`

Reply:
Changed. **NOTE:** this is why there are now changes to compression and the legacy pool in this PR.